src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs

```diff
@@ -0,0 +1,837 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Xml;
+
+using Internal.IL;
+using Internal.IL.Stubs;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+using Debug = System.Diagnostics.Debug;
+
+namespace ILCompiler
+{
+    public class FeatureSwitchManager : ILProvider
+    {
+        private readonly FeatureSwitchHashtable _hashtable;
+        private readonly ILProvider _nestedILProvider;
+
+        public FeatureSwitchManager(ILProvider nestedILProvider, IEnumerable<KeyValuePair<string, bool>> switchValues)
+        {
+            _nestedILProvider = nestedILProvider;
+            _hashtable = new FeatureSwitchHashtable(new Dictionary<string, bool>(switchValues));
+        }
+
+        private BodySubstitution GetSubstitution(MethodDesc method)
+        {
+            if (method.GetTypicalMethodDefinition() is EcmaMethod ecmaMethod)
+            {
+                AssemblyFeatureInfo info = _hashtable.GetOrCreateValue(ecmaMethod.Module);
+                if (info.BodySubstitutions != null && info.BodySubstitutions.TryGetValue(ecmaMethod, out BodySubstitution result))
+                    return result;
+            }
+
+            return null;
+        }
+
+        private object GetSubstitution(FieldDesc field)
+        {
+            if (field.GetTypicalFieldDefinition() is EcmaField ecmaField)
+            {
+                AssemblyFeatureInfo info = _hashtable.GetOrCreateValue(ecmaField.Module);
+                if (info.BodySubstitutions != null && info.FieldSubstitutions.TryGetValue(ecmaField, out object result))
+                    return result;
+            }
+
+            return null;
+        }
+
+        public bool HasSubstitutedBody(MethodDesc method)
+        {
+            return GetSubstitution(method) != null;
+        }
+
+        public bool HasSubstitutedValue(FieldDesc field)
+        {
+            return GetSubstitution(field) != null;
+        }
+
+        public override MethodIL GetMethodIL(MethodDesc method)
+        {
+            BodySubstitution substitution = GetSubstitution(method);
+            if (substitution != null)
+            {
+                return substitution.EmitIL(method);
+            }
+
+            // BEGIN TEMPORARY WORKAROUND
+            //
+            // The following lines should just be:
+            // return _nestedILProvider.GetMethodIL(method);
+            // But we want to allow this to be used as a general-purpose IL provider.
+            //
+            // Rewriting all IL has compilation throughput hit we don't want.
```

Comment:
How much throughput hit does this change have? (Trying to see whether we need to bother with rewriting only once we detect there is something to rewrite.)

Reply:
I'll measure. It's possible it's pretty cheap. We run this in the multithreaded phase.

Reply:
I'm getting pretty noisy results (+- 2%) and this is basically within the noise)